File: ecmascript/anagram/anagram.js
export default class Anagram {
  constructor(word) {
    this.word = word;
  }
  matches(...args) {
    let wordList = (args[0].constructor === Array) ? args[0] : args;
    return wordList.filter((word) => {
      return (this.word.toLowerCase() !== word.toLowerCase()) &&
             (this.word.length === word.length) &&
             (this.sort(this.word) === this.sort(word));
    });
  }
  sort(string){
    return string.toLowerCase().split("").sort().join("");
  }
}

File: elixir/anagram/anagram.exs
defmodule Anagram do
  def match(base, candidates) do
    candidates
    |> Enum.filter(fn(word) ->
      String.downcase(word) != String.downcase(base) &&
      String.length(word) == String.length(base) &&
      sort_and_glue(word) == sort_and_glue(base)
    end)
  end

  def sort_and_glue(str) do
    str
    |> String.downcase
    |> String.codepoints
    |> Enum.sort
    |> Enum.join
  end
end
